src/lib/transition_router/transition_router.ml

```diff
@@ -93,34 +92,35 @@ let start_transition_frontier_controller ~context:(module Context : CONTEXT)
           ?valid_cb ~pipe_name:name ~logger )
       ()
   in
-  transition_reader_ref := transition_frontier_controller_reader ;
-  transition_writer_ref := transition_frontier_controller_writer ;
+  let transition_writer_ref =
```

Comment:
Arguably change of semantics behind `transition_writer_ref` is the most non-trivial change of the PR.

I strongly advise to extract this change as a separate commit.

Also, I think there is a bug here.

Original intention seemed to have been to remove the need to pass `transition_reader_ref` everywhere (because where it's used what's inside is always a reader created locally, hence we don't need a `ref` to use the correct value). And also to remove redundant pipe creation to start initialization (a pipe that gets discarded immediately). For this reason the `transition_writer_ref` was made an optional argument (and no argument in case of `initialize` function).

So the code below does it right that it uses `ref transition_frontier_controller_writer` as the default value. It does it wrong in not setting ref to `transition_frontier_controller_writer` when `transition_writer_ref` is not `None`.

So the code I think should be here is `Option.value_map ~f:(fun r -> r := transition_frontier_controller_writer; r) transition_writer_ref
    ~default:(ref transition_frontier_controller_writer)`

And same in `start_bootstrap_controller`.

Reply:
Isn't `Option.value_map ~f:(fun r -> r := transition_frontier_controller_writer; r) transition_writer_ref ~default:(ref transition_frontier_controller_writer)` the same as `ref transition_frontier_controller_writer` ?

I think I’m missing something because to me this makes the `?transition_writer_ref` argument given to the `start_transition_frontier_controller` function useless, as it is just replaced with `transition_frontier_controller_writer`

Reply:
A very good question!

When `?transition_writer_ref:None` is passed, following codes are identical by definition:

```ocaml
(* code 1 *)
Option.value_map ~f:(fun r -> r := transition_frontier_controller_writer; r) transition_writer_ref ~default:(ref transition_frontier_controller_writer)

(* code 2 *)
ref transition_frontier_controller_writer
```

However when `?transition_writer_ref:(Some myref)` is passed, behavior changes. Code that calls `start_transition_frontier_controller` may use `myref` in some other thread. In particular we use it the following way:

```ocaml
don't_wait_for
@@ Strict_pipe.Reader.iter_without_pushback valid_transition_reader2
  ~f:(fun (`Block enveloped_transition, `Valid_cb vc) ->
   don't_wait_for @@
   let%map () = (* some code removed for clarity *) in
   Strict_pipe.Writer.write !transition_writer_ref
      (`Block enveloped_transition, `Valid_cb (Some vc)) ) )
```

We have an infinite loop of `start_transition_frontier_controller` transitioning to `start_bootstrap_controller` and then back, depending on state of sync. However the async thread above is started once with some ref. It's important that we write to **that same** variable instead of creating a new one as the code `let transition_writer_ref = ref transition_frontier_controller_writer` would do.

Reply:
Thanks a lot for this explaination, it makes more sense now.

I made a separate commit including your new version : https://github.com/MinaProtocol/mina/pull/16145/commits/0cbd9472709757ab3e80adb5ebc73bd79723b42a